### PR TITLE
Use `push` for 8/12 byte struct args on x86

### PR DESCRIPTION
**Pull request by SingleAccretion**
It is the case that currently for code like the following:
```cs
private static void Problem(StructWithIndex* p)
{
    CallForStructWithIndex(*p);
}

struct StructWithIndex
{
    public int Index;
    public int Value;
}
```
On x86 we generate the following:
```asm
       vmovq    xmm0, qword ptr [ecx]
       sub      esp, 8
       vmovq    qword ptr [esp], xmm0
       call     RyuJitReproduction.Program:CallForStructWithIndex(StructWithIndex)
```
While could generate:
```asm
       push     dword ptr [ecx+4]
       push     dword ptr [ecx]
```
Which is both a little faster (ah-hoc benchmarking on my machine shows ~10% improvement) and smaller. This change does just that, using `push [addr]` for 8 and 12 byte structs (the latter made the code simple).

To achieve that, some code refactoring was undertaken: we already had a path that did what we needed, but it was only used for structs with GC pointers. I deleted the double meaning of `Kind::Push` - for `TYP_STRUCT` it now means only one thing, that is `genStructPutArgPush`, while before this change it could also go down the "unroll" path (for the very specific cases of 8 and 12 byte structs that we are moving to using `push`es).

As a cleanup item, `PartialRepInstr` kind was added, to represent the algorithm we use for storing structs with GC pointers on x64 Unix.

There will be two types of diffs with this change: first the one showed above, second - removal of unnecessary `vzeroupper` instructions that were being generated because LSRA was allocating an unnecessary XMM register for `Kind::Push` `PUTARG_STK`s with GC pointers.

[Diffs](https://github.com/SingleAccretion/diffs-repository/blob/main/runtime-65387/win-x86.md) - x86 only ([no](https://dev.azure.com/dnceng/public/_build/results?buildId=1614490&view=ms.vss-build-web.run-extensions-tab) diffs on other targets as expected). All regressions are alignment artifacts.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
Not sure what's up with SPMI in CI. Will assume the Windows x64 timeout is not related (it does not use `PUTARG_STK`).

@dotnet/jit-contrib some nice CQ improvements for x86 (as always, because I need them for another change).

**Comment by kunalspathak** [comment]
From [unix-arm logs](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-65387-merge-e64c547067ce446eaf/unix-arm/1/console.a2dab829.log?sv=2019-07-07&se=2022-03-07T22%3A31%3A53Z&sr=c&sp=rl&sig=QwG5Pfjftw7O7nlIjezSVEKussw1Ow36Lx%2B6vSSs42A%3D), seems some errors:

```
ERROR: Unexpected exception was thrown.

ERROR: Unexpected exception was thrown.

ERROR: Method 354 of size 7557059 failed to load and compile correctly by JIT2.

ERROR: main method 354 of size 7557059 failed to load and compile correctly.

ERROR: Couldn't load base metrics summary created by child process
```

Can you double check on your local machine?

For x86, they likely have timed out while producing .dasm files.

File: src/coreclr/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -3453,18 +3409,180 @@ void CodeGen::genStructPutArgUnroll(GenTreePutArgStk* putArgNode)
...
+    {
+        assert(srcAddr->OperIsLocalAddr());
+
+        srcLclNum    = srcAddr->AsLclVarCommon()->GetLclNum();
+        srcLclOffset = srcAddr->AsLclVarCommon()->GetLclOffs();
+    }
+
+    ClassLayout*   layout   = src->AsObj()->GetLayout();
+    const unsigned byteSize = putArgNode->GetStackByteSize();
+    assert(byteSize % TARGET_POINTER_SIZE == 0);
```
Comment (kunalspathak) [inline src/coreclr/jit/codegenxarch.cpp:3469]:
```suggestion
    assert((byteSize % TARGET_POINTER_SIZE == 0) && (byteSize  < XMM_REGSIZE_BYTES));
```

**Review by kunalspathak — COMMENTED** [review]
Nice refactoring. Changes LGTM except the failure on superpmi linux-arm that you want to double check.

**Comment by SingleAccretion** [comment]
> Can you double check on your local machine?

@kunalspathak Thank you for pointing that failure out, I did unfortunately miss it in the logs. It does not seem related: no ARM code paths were modified, and I can reproduce it both with and without this change.

It sure does look curious though: we're OOMing the Jit (2GB+ memory consumption and then an error in the CRT is hit).

> Could you paste the sample diff?

They're all are like this:
```diff
        sub      esp, 20
-       vzeroupper
        xor      eax, eax
```
I suppose the better wording for this would have been that "LSRA was over-reporting the use of XMM registers".

File: src/coreclr/jit/codegenxarch.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/codegenxarch.cpp:3469]: Ah right. Is there a way to check something like `(either it is a structs with GC pointers) or (bytesize check that I suggested)`.

**Comment by kunalspathak** [comment]
> It sure does look curious though: we're OOMing the Jit (2GB+ memory consumption and then an error in the CRT is hit).

Where do you see that? in the logs?

File: src/coreclr/jit/codegenxarch.cpp (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/jit/codegenxarch.cpp:3469]: That is certainly possible. Will add.

**Comment by SingleAccretion** [comment]
> Where do you see that? in the logs?

No, from locally replaying the `coreclr_tests.pmi.Linux.arm.checked` collection (no `.mc` because SPMI itself crashes).

**Comment by SingleAccretion** [comment]
As I note, yes.

**Comment by SingleAccretion** [comment]
Categorizing fuzzer failures.

It seems Fuzzlyn is timing out pretty consistently on Linux ARM, and we have one failure there:
```
Found example with seed 16498413861848824582 that hits error
JIT assert failed:
Assertion failed 'inVarToRegMap[varIndex] == REG_STK' in 'S1:M50(int):long:this' during 'LSRA allocate' (IL size 4270)

File: /__w/1/s/src/coreclr/jit/lsra.cpp Line: 4050
```
That's pre-existing - seen [in an earlier run](https://dev.azure.com/dnceng/public/_build/results?buildId=1599668&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=96d68eba-7198-517e-3083-eb7169f4ccb9) ([logs link](https://helix.dot.net/api/jobs/f1ea5d8a-329f-490a-93a5-087926001070/workitems/Partition3?api-version=2019-06-17)).

Antigen on different targets has quite a few "uncategorized" issues (guessing these are for "wrong result" cases?), but the only x86 failure looks like #64787, and is an inlining assert (that this backend-only change could not have caused, I suppose).

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Comment by kunalspathak** [comment]
@SingleAccretion - Keep an eye on the [superpmi-replay run](https://dev.azure.com/dnceng/public/_build?definitionId=1055&_a=summary) that will kick off on your commit to make sure there are no errors.

**Comment by kunalspathak** [comment]
Improvements - https://github.com/dotnet/perf-autofiling-issues/issues/3651

**Pull request merged**